atom/browser/net/resolve_proxy_helper.cc

```diff
@@ -4,76 +4,79 @@
+    : binding_(this), browser_context_(browser_context) {}
 
 ResolveProxyHelper::~ResolveProxyHelper() {
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+  DCHECK(!owned_self_);
+  DCHECK(!binding_.is_bound());
   // Clear all pending requests if the ProxyService is still alive.
   pending_requests_.clear();
 }
 
 void ResolveProxyHelper::ResolveProxy(const GURL& url,
                                       const ResolveProxyCallback& callback) {
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
   // Enqueue the pending request.
   pending_requests_.push_back(PendingRequest(url, callback));
 
   // If nothing is in progress, start.
-  if (pending_requests_.size() == 1)
+  if (!binding_.is_bound()) {
+    DCHECK_EQ(1u, pending_requests_.size());
     StartPendingRequest();
+  }
 }
 
 void ResolveProxyHelper::StartPendingRequest() {
-  auto& pending_request = pending_requests_.front();
-  context_getter_->GetNetworkTaskRunner()->PostTask(
-      FROM_HERE, base::BindOnce(&ResolveProxyHelper::StartPendingRequestInIO,
-                                base::Unretained(this), pending_request.url));
-}
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+  DCHECK(!binding_.is_bound());
+  DCHECK(!pending_requests_.empty());
 
-void ResolveProxyHelper::StartPendingRequestInIO(const GURL& url) {
-  auto* proxy_service =
-      context_getter_->GetURLRequestContext()->proxy_resolution_service();
   // Start the request.
-  int result = proxy_service->ResolveProxy(
-      url, std::string(), &proxy_info_,
-      base::Bind(&ResolveProxyHelper::OnProxyResolveComplete,
-                 base::RetainedRef(this)),
-      nullptr, nullptr, net::NetLogWithSource());
-  // Completed synchronously.
-  if (result != net::ERR_IO_PENDING)
-    OnProxyResolveComplete(result);
+  network::mojom::ProxyLookupClientPtr proxy_lookup_client;
+  binding_.Bind(mojo::MakeRequest(&proxy_lookup_client));
+  binding_.set_connection_error_handler(
+      base::BindOnce(&ResolveProxyHelper::OnProxyLookupComplete,
+                     base::Unretained(this), base::nullopt));
+  content::BrowserContext::GetDefaultStoragePartition(browser_context_)
+      ->GetNetworkContext()
+      ->LookUpProxyForURL(pending_requests_.front().url,
+                          std::move(proxy_lookup_client));
 }
 
-void ResolveProxyHelper::OnProxyResolveComplete(int result) {
+void ResolveProxyHelper::OnProxyLookupComplete(
+    const base::Optional<net::ProxyInfo>& proxy_info) {
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
   DCHECK(!pending_requests_.empty());
 
-  std::string proxy;
-  if (result == net::OK)
-    proxy = proxy_info_.ToPacString();
+  binding_.Close();
 
-  original_thread_->PostTask(
-      FROM_HERE, base::BindOnce(&ResolveProxyHelper::SendProxyResult,
-                                base::RetainedRef(this), proxy));
-}
+  // Clear the current (completed) request.
+  PendingRequest completed_request = std::move(pending_requests_.front());
```

Comment:
Is `std::move` needed here? How can it be `move`d out of `pending_requests_` but also require a `pop_front()`?

Reply:
`std::move` cancels the reference returned by `front` and `pop_front` will invalidate the iterator and references from the container.